Daemonset deployment mode

Pacho20 opened the pull request:
Introduce support for DaemonSet-based deployment mode for clusters running without the Machine Config Operator (MCO).

This feature enables an alternative deployment path by detecting the absence of the MachineConfig CRD and activating DaemonSet mode via a dedicated feature gate. It includes logic for deploying and managing DaemonSets on targeted nodes, using labels to control execution and track installation status.

There are still many TODOs in the implementation, and the list will be updated as development progresses.

spotlesstofu commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,451 @@
+		r.DeploymentMode = DaemonSet
+	} else {
+		r.Log.Info("Deployment mode will be set to MachineConfig")
+		r.DeploymentMode = MachineConfig
```
What's the case for explicitly setting the DeploymentMode to MachineConfig?

Leaving it untouched may help avoid clashes with future reconcile code outside of this feature.

spotlesstofu commented:
/ok-to-test

spotlesstofu commented:
/ok-to-test

spotlesstofu commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,1014 @@
+					Labels: dsLabelSelectors,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: "default", // TODO: Which service account should be used?
```
I see we already use this ServiceAccount in the operator. If it works, I'm OK with it.

spotlesstofu commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,1014 @@
+	}
+
+	return imageString, nil
+}
```
`GetExtensionImage` and `GetCliImage` look the same. Can deduplicate?

Pacho20 replied:
Done

spotlesstofu commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,1014 @@
+    set_status "waiting_for_reboot"
+}
+
+# Initial wait: avoid doing anything if a previous staged update is pending
```
It may happen that an update is staged for other packages. It would be nice to check if kata is installed before requesting a reboot, and installing the package if it's not.

gcoon151 replied:
good suggestion. Before we start trying to deploy kata check to see if there's pending packages. We'll get stuck waiting on something that has nothing to do with kata.

Pacho20 replied:
Modified the script to also check the package list.

spotlesstofu reviewed (changes requested):
Thanks Patrik for working on this new feature.

I added a couple of inline comments about the logic.

Please move the shell scripts to a separate file and include them, or mount them as an immutable ConfigMap.

Nice to have: deduplicate code.

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,90 @@
+// If the MachineConfig Add-on is unavailable and the FeatureGateState is Enabled,
+// the deployment mode is set to DaemonSet. Otherwise, it defaults to MachineConfig.
+// If the FeatureGateState is Disabled, the deployment mode remains MachineConfig,
+// regardless of the MachineConfig Add-on's availability.
```
Have you considered inverting the logic? For example: if the MachineConfig add-on is available and the feature gate is disabled, then we use MachineConfig (the current behavior); otherwise, if the feature gate is enabled, we always use the new DaemonSet mode, regardless of MC availability.

Just wondering if we really need to check MC availability when we're explicitly enabling DaemonSet mode.

spotlesstofu replied:
This is more of a future-looking design choice.

In a scenario where the feature comes enabled by default, it makes sense to also check for MachineConfig availability and select the deployment mode accordingly.

This logic conveys that the MachineConfig is a more mature mechanism and should be preferred when it's available.

beraldoleal replied:
How would a user force the use of DaemonSet, regardless of whether the MC CRD is present or not?

Maybe I'm misinterpreting the intent here, but if this is just meant as a fallback, then I’d suggest renaming the feature to something like `DaemonSetDeploymentFallback` to avoid giving the impression that users are explicitly enabling DaemonSet mode.

Pacho20 replied:
If the MCO manages the nodes, we don’t want to interfere with that. Since we’re altering runtime configurations, there’s a risk that the DaemonSet could conflict with the MachineConfig.

Also, I don’t see a use case for choosing the DaemonSet installation over the MCO when the MCO is already being used in the cluster.

beraldoleal replied:
Fair point about avoiding MCO conflicts. But if the feature is called “enable DaemonSet mode,” users will expect it to override MCO. Relying on docs isn’t enough. The name should make the intent clear on its own. So I’d suggest renaming it to avoid confusion.

bpradipt replied:
We may have a use case for forcing Daemonset mode - during testing, installing custom artefacts for dev-preview which cannot be installed via MC. Why not have a force/override mode as an option for the feature?

gcoon151 replied:
Just want to mention again that hypershift is the future on places like Azure and in hosted control planes (cloud providers) the masters have MCO so I think that will break the today code expectations of what works and what doesn't. I'm not sure if that influences how we want to handle this.

Pacho20 replied:
I modified the feature gate logic in commit c7afd44892de1be200793faf93809b6a7ceae6c9.

The feature gate now supports three values:
- `MachineConfig` (default)
- `DaemonSet`
- `DaemonSetFallback`

Use `DaemonSet` to explicitly force DaemonSet mode.

beraldoleal replied:
Closing this thread, since @Pacho20  already implemented the modes. Feel free to reopen if needed.

beraldoleal commented on controllers/fg_handler.go:
```diff
@@ -12,11 +12,13 @@ const (
 var DefaultFeatureGates = map[string]bool{
 	ConfidentialFeatureGate: false,
 	LayeredImageDeployment:  false,
+	DaemonSetDeployment:     false,
```
Just to give you some visibility, there is a discussion going on that we are planning to rename "Feature Gates" to just "Config Option":

https://github.com/openshift/sandboxed-containers-operator/pull/827

So we are planing renaming the config map name, soon. Giving this is a new feature, I'm concerned the users will have to learn once and the change it latter.

@gkurz, @bpradipt ... maybe we could rename the config map as soon as possible, so this new config option would be already at the right place?

beraldoleal replied:
We discussed this today in a call, and we fine with changing this the future as it will be TP. 

Closing this thread, feel free to reopen if needed.

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -266,8 +269,58 @@ func (r *KataConfigOpenShiftReconciler) addPeerPodsConfigDaemonSet() error {
 }
 
 func (r *KataConfigOpenShiftReconciler) GetExtensionImage() (string, error) {
```
Please, could you squash this commit with the previous one? This way, when reviewing individual commits we don't need to follow your flow.

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -203,6 +203,24 @@ func (r *KataConfigOpenShiftReconciler) processDaemonSetKataConfigInstallRequest
 		}
```
Same here, squash with the 1e96e993911a3362c65aa2e38b7cccd6de899f0f

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -349,17 +349,24 @@ func (r *KataConfigOpenShiftReconciler) DaemonSetForKataInstall(imageString stri
 		kataInstallDsName       = "osc-rpm-install"
 
 		script = `
-sleep infinity
```
Same here, please squash with 1e96e993911a3362c65aa2e38b7cccd6de899f0f

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -348,8 +348,32 @@ func (r *KataConfigOpenShiftReconciler) DaemonSetForKataInstall(imageString stri
 		nodeSelector            = r.getNodeSelectorAsMap()
```
Same here, please squash with 1e96e993911a3362c65aa2e38b7cccd6de899f0f.

beraldoleal reviewed (commented):
Hey @Pacho20, thanks for this. Appreciated.

I left a few comments for you, but my first suggestion here, is to squash the commits that are refactoring and changing the same code base, so the commit-by-commit review would be easier.

gcoon151 commented on config/peerpods/osc-configs-for-ds.yaml:
```diff
@@ -0,0 +1,585 @@
+  name: osc-configs
+  namespace: openshift-sandboxed-containers-operator
+#immutable: true
+data:
```
Should the installscript, 50-kata-remote, and configuration-remote.toml all move to their own configmaps so they don't have yaml inside yaml problems. I think I see a TODO comment that you want to do that though.

gcoon151 replied:
I see you moved these. This is handled. Thanks @Pacho20 !

gcoon151 commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,1014 @@
+							Image:           imageString,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							SecurityContext: &corev1.SecurityContext{
+								// TODO: do we really need to run as root?
```
The answer here is yes. We're installing things on workers.

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -81,9 +80,3 @@ func (r *KataConfigOpenShiftReconciler) isMCAvailable() (bool, error) {
 }
-
-
-func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, error) {
```
This part of the commit should be squashed with 2c3da8db78973bba3f1479d551857690b9b6b65a

Pacho20 replied:
I squashed that part—hope it's okay this way

littlejawa commented:
/ok-to-test

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,122 @@
+)
+
+// Create enum to represent the state of the deployment mode
+type DeploymentMode int
```
Since this file is called `daemonset_handler`, I feel most of the "DeploymentMode logic" isn't specific to the daemonset handler. So maybe move this out of this file?

Pacho20 replied:
Valid point. It was originally called `daemonset_handler` because it handled the DaemonSet Feature Gate. Since that's no longer the case, I'll rename the file. The entire logic now relates to DeploymentMode.

beraldoleal commented on controllers/fg_handler.go:
```diff
@@ -40,22 +45,35 @@ const (
+			if err != nil {
+				return nil, err
+			}
+			fgStatus.DeploymentModeOption = mode
```
I agree with the new logic, but since it changes behavior (invalid values now error instead of falling back to defaults, for both bools and the custom type), I think we should keep the old semantics for this z-release and defer the stricter validation to a minor/major bump.

Pacho20 replied:
Oh, you're right. I'll modify the code to preserve the old semantics.

beraldoleal commented on controllers/daemonset_handler.go:
```diff
@@ -0,0 +1,122 @@
+//   - If the mode is DaemonSetOption, the deployment mode is forcibly set to DaemonSet, regardless of MC availability.
+//   - If the mode is DaemonSetFallbackOption, the deployment mode is set to DaemonSet only if the MachineConfig Add-on is unavailable.
+//     Otherwise, it defaults to MachineConfig.
+func (r *KataConfigOpenShiftReconciler) handleDaemonSetFeature(mode DeploymentModeOption) error {
```
Now that this method is dealing with multiple modes, the method name is a bit misleading... I would suggest something like `resolveDeploymentMode` or `updateDeploymentMode`.

Pacho20 replied:
True, I'll rename it. I’m thinking of using something like `handleDeploymentModeFeature` to follow the naming convention of the feature gate handlers.

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -146,6 +148,9 @@ func (r *KataConfigOpenShiftReconciler) Reconcile(ctx context.Context, req ctrl.
 
+	if r.DeploymentMode == DaemonSetMode {
+	}
+
```
leftover?

beraldoleal replied:
Oh, now that I'm reviewing 15c66c431323de800732efa7b66d29e3c6f05895 I see. You added an empty if on the first commit, and then added a return in a later commit. This needs to be squashed.

beraldoleal commented on controllers/daemonset_reconcile.go:
```diff
@@ -0,0 +1,1145 @@
+package controllers
```
I'm concerned about this duplication here, and how we will keep maintaining both of them for a while.

Any reason for this?

Pacho20 replied:
Just to clarify, when you mention duplication, are you referring to the separate reconciliation logic paths, or something else?

beraldoleal replied:
Yes, I'm talking about the "reconcile loop".

Pacho20 replied:
The logic of the reconcile function is deeply coupled with the MachineConfig CRD. It includes numerous checks and MachineConfig-specific operations. At the beginning of development, I chose to implement a separate reconcile loop to minimize changes to the existing codebase and avoid unintended side effects. I had no clear idea how the new implementation would look, but given the tight coupling, I assumed there wouldn’t be much overlap and that integration would be difficult.
Because of that, I didn’t attempt to merge the two reconcile loops early on. However, over the past few days, I’ve been exploring ways to make the solution more maintainable, since I agree that having two separate flows isn’t ideal. I just pushed some changes that merge parts of the logic and remove some duplication—these were relatively straightforward and safe to refactor. There are now shared function calls between the MachineConfig and DaemonSet deployment modes.
That said, the core logic still remains distinct, and there are conditional checks that determine which major function to call. There’s definitely room for further improvement. But tighter integration would require a deeper refactor and a better understanding of the original design. That’s not something I feel comfortable doing alone, given the complexity and potential impact. I’m also not sure if that level of integration is something we can—or should—tackle in this initial release.

beraldoleal replied:
Please vide my other commit, avoid adding new commits if you are changing the logic on previous commits you introduced.

beraldoleal replied:
I just saw your latest changes, and I'm happy with the split into methods and the reuse for both reconciles. It reduces a lot the duplication. And also, break our existing reconcile loop into smaller methods. Awesome.

I'm happy with this, so I'm closing this thread, @bpradipt feel free to reopen if you still see some room for improvements.

beraldoleal reviewed (commented):
Hi @Pacho20, thanks for keeping pushing this PR. Appreciated. 

I left a few more comments for you, but also lets see if others join the party.

beraldoleal commented:
Also, please rebase, since go.sum got a conflict.

bpradipt commented on controllers/daemonset_reconcile.go:
```diff
@@ -0,0 +1,1086 @@
+// The DaemonSet's Pod contains three containers: one installs the binaries, one modifies the node labels, and one sets the runtime log level.
+func (r *KataConfigOpenShiftReconciler) daemonSetForKataInstall(action KataDsAction) (*appsv1.DaemonSet, error) {
+	// This image contains the binaries for kata-containers
+	extensionImageString, err := r.GetExtensionImage()
```
Wouldn't it make sense to build a single container image with deployment artefacts?
- kata binaries
- CLI
- scripts
- ..

This way we don't need to create additional configMap with embedded script? Also will reduce to effort w.r.to setting up separate build pipelines.

gcoon151 replied:
@Pacho20 is new to IBM and openshift. Which image would you want to put these in? Also would we be able to make a new OCP image in time to release all this for September? We may need some help determining where to do this @bpradipt

bpradipt replied:
I'm trying to understand this whole code as this is the key piece of creating the DS to install.
My understanding is that this code is fetching the ExtensionImage and the CLI image and using an install script in the config yaml for installation.

The approach is creative in a way that you rely on images already present (extension and CLI). 
Personally, using embedded script in yaml is something I would avoid. It makes it hard to maintain and yamls are brittle. 

I think a better approach is to create a separate DS image having the Kata rpm, required CLI, config tomls and install script and use this for installation. 

And my reasoning is that it helps us to use static scanning tool for the script which can be removed later with compiled binary and the ability to replace the daemonset image for different use cases. 

Now creating a new image is an extra work w.r.to setting up Konflux pipelines. 

Also there could be difference of opinion here w.r.to creating a separate image vs the existing method.

Let's hear from @beraldoleal @gkurz @littlejawa and others and decide accordingly.

gcoon151 replied:
Appreciate the explanation and will await further review from the others. The approach we took was specifically because we have no method of publishing an image with the script we need or modifying the openshift extensions image to have it. Very open to doing this however everyone feels is appropriate.

beraldoleal replied:
We should stay aligned with how we deploy across the project to avoid mixing methods. Since [there’s already an effort to move the crio/kata config into the kata RPM](https://github.com/openshift/sandboxed-containers-operator/pull/941), this DS could follow the same path, otherwise it risks becoming a separate “thing” that’s harder to maintain and ship in future versions. I’d also avoid introducing extra config maps, even if used only internally, since they tend to add confusion.

Now, about the scripts, we already have `scripts/install-helpers/`, so the needed files could go there, and the image could consume them at build time as other images do.

If we have time to deploy this image in our pipeline, that’s the option I’d vote for (unless we have a better alternative).

@littlejawa, how hard do you think it would be to add this new component/image in Konflux? From what I see, hermetic builds shouldn’t be a problem, but do you see any other enterprise contract breaking here?

littlejawa replied:
It's not that hard to create and release a new image as part of our operator, as long as you can build it with a regular Dockerfile.
I think we're talking about a relatively simple image here? Not requiring special access rights, like dm-verity?

Is this image going to ship the RPM? I thought the install script was installing from repositories?

beraldoleal replied:
> Is this image going to ship the RPM? I thought the install script was installing from repositories?

No, the idea is that the RPM would be fetched from the extension the same way we do today. The image would have only the scripts and some artifacts.

bpradipt commented on controllers/daemonset_reconcile.go:
```diff
@@ -0,0 +1,1086 @@
+								Privileged: &runPrivileged,
+								RunAsUser:  &runAsUser,
+							},
+							Command: []string{"/bin/bash", "/scripts/osc-configs-script.sh"},
```
See the comment https://github.com/openshift/sandboxed-containers-operator/pull/825/files#r2313413123

bpradipt commented on controllers/daemonset_reconcile.go:
```diff
@@ -0,0 +1,1086 @@
+//
+// It returns false if no ConfigMap was created due to an error or
+// if the existing ConfigMap is identical to the provided YAML data.
+func (r *KataConfigOpenShiftReconciler) addImmutableConfigMap(path, yaml string) (bool, error) {
```
Is there a benefit of using the script from the configmap vs adding the script in the container image and run it through the usual static checks etc ?

gcoon151 replied:
I had suggested this from experience that yaml inside yaml can often have issues in ocp. The line wraps start to be interpreted by ocp when they should be running it verbatim inside the container.

beraldoleal replied:
Closing this as it was removed from this PR.

beraldoleal replied:
Wrong, thread, reopening.

bpradipt commented on controllers/openshift_controller.go:
```diff
@@ -488,6 +513,54 @@ func (r *KataConfigOpenShiftReconciler) processDaemonsetForCAA() *appsv1.DaemonS
 
+// Handles provider specific parts of the CAA Ds
+// Modifies the DaemonSet if needed
+func (r *KataConfigOpenShiftReconciler) processProviderConfigCAA(ds *appsv1.DaemonSet) error {
```
I don't think this is related to this PR of daemonset deployment mode or am I missing something ?

gcoon151 replied:
The reason we're building the daemonset code is to make it run on ROKS on IBM cloud which doesn't have MCO (it runs hypershift). This was something else that needed to be fixed to make that work. We can move to a different PR if desired, but this is a part of the whole thing we're wanting to make work in 1.10.2. Your call.

beraldoleal replied:
Closing this as it was removed from this PR.

bpradipt commented on controllers/image_generator.go:
```diff
@@ -54,12 +54,14 @@ const (
 	peerpodsCMAzureImageKey       = "AZURE_IMAGE_ID"
 	peerpodsCMGCPImageKey         = "PODVM_IMAGE_NAME"
 	peerpodsLibvirtImageKey       = "LIBVIRT_IMAGE_ID"
+	peerpodsCMIBMCloudImageKey    = "IBMCLOUD_PODVM_IMAGE_ID"
```
You can extract this code into a separate PR and get it merged earlier.
This was we will be focussing only on the DS deployment mode in this PR

gcoon151 replied:
Same as earlier comment. We can move to different PR but that will be a part of what we require and need in 1.10.2 by end of September. Your call @bpradipt

gkurz replied:
I agree with @bpradipt. We will make sure that all required changes are there for 1.10.2.

The workflow will be :
- submit changes to the devel branch
- backport (cherry-pick) to the osc-release-v1.10 branch

beraldoleal replied:
I agree with @bpradipt as well. I would say its better to split. Both can still target 1.10.2, but separate PRs simplify review and make those adjustments easier.

beraldoleal replied:
closing this thread, since the new provider is part of #950

gkurz commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,122 @@
+	r.Log.Info("Feature gate", "featuregate", DeploymentModeConfig, "state", mode)
+
+	if mode == MachineConfigOption {
+		r.Log.Info("Deployment mode will be set to MachineConfig")
```
Availability of the MachineConfig CRD should also be checked in this case IMO. It is preferred to detect early that the required config won't work over hitting an error later in the MCO-based reconcile logic.

gcoon151 replied:
@gkurz we think we resolved this. Can you confirm?

gkurz commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,122 @@
+
+	// Attempt to GET any MachineConfig to verify if the GVK is known.
+	// If the CRD isn’t installed, it will return a NoMatchError.
+	err := r.Client.Get(context.Background(), client.ObjectKey{Name: machineConfigCrdName}, mc)
```
It is rather confusing that you look for a MachineConfig object named like the MachineConfig CRD... Please use some dummy name instead.

beraldoleal replied:
@gkurz, this was fixed, I'm closing the thread, but feel free to reopen if needed.

gkurz commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,122 @@
+		return false, err
+	}
+
+	return true, nil
```
MachineConfig CRD is also present in this case but it won't print the message... :-\

What about simplifying the return branches ? Something like :

```
	err := r.Client.Get(context.Background(), client.ObjectKey{Name: "foo"}, mc)
	if err == nil || k8serrors.IsNotFound(err) {
		r.Log.Info("MachineConfig CRD is present")
		return true, nil
	}

	if meta.IsNoMatchError(err) {
		r.Log.Info("MachineConfig CRD not found")
		return false, nil
	}

	return false, err
```

gcoon151 replied:
@gkurz we think we resolved this. Can you confirm?

gkurz reviewed (commented):
Hi @Pacho20 !

Thanks for submitting this. I could only partially review the first commit so far, I'll try to devote some time everyday to make progress. Anyway, I've already some comments to share. PTAL.

c3d commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,122 @@
+
+	mcAvailable, err := r.isMCAvailable()
+	if err != nil {
+		r.Log.Info("Error checking if MC is available")
```
```suggestion
		r.Log.Info("Error checking if MachineConfig is available")
```

beraldoleal replied:
Done.

c3d commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	if mode == DaemonSetFallbackOption && !mcAvailable {
+		r.Log.Info("MC is not available, deployment mode will be set to DaemonSet")
```
```suggestion
		r.Log.Info("MachineConfig is not available, deployment mode will be set to DaemonSet")
```

Pacho20 replied:
@gkurz this should now be resolved. Can you confirm?

gkurz replied:
AFAICT the log still has `MC`.

Pacho20 replied:
Sorry, I don’t know how I missed that. It should be good now.

gkurz commented on controllers/daemonset_reconcile.go:
```diff
@@ -89,20 +99,20 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
-			return ctrl.Result{}, updateErr
-		}
-		return res, err
+			res, err := r.processKataConfigDeleteRequestDaemonSet()
```
Space damage. Be sure to run `make fmt` for each individual commit. It is also a good practice to run `make vet` as well.

gcoon151 replied:
I think was fixed when it moved to here https://github.com/openshift/sandboxed-containers-operator/pull/950 the commit doesn't exist that has this error now.

c3d commented on controllers/fg_handler.go:
```diff
@@ -40,22 +45,38 @@ const (
 	if err == nil {
-		for feature, value := range cfgMap.Data {
-			fgStatus.FeatureGates[feature] = value == "true"
+		if value, ok := cfgMap.Data[ConfidentialFeatureGate]; ok {
```
Changing the `FeatureGateStatus` from containing a map to containing explicit values is what forces you to copy-paste the same code again and again here. This scales OK with three fields, but if we add more, that function will need to be revisited every time. Do we want that?

beraldoleal replied:
@c3d, right, we asked him to add DeploymentModeOption over Slack to cover what the code was already doing and to support the "force mode". Since we don’t have just bools anymore, the extra code here is expected.

We’re also [moving from feature gates to config options](https://github.com/openshift/sandboxed-containers-operator/blob/devel/docs/CONFIG-OPTIONS.md). I don’t love the duplication either, but for now I’d stick with this. If we add more gates later, we can always revisit with a table-driven/helper pattern.

beraldoleal replied:
I will close this thread, feel free to reopen if we came up with a better idea.

gkurz replied:
Such a change should really happen in a preceding commit.

Pacho20 replied:
I've created a new commit. Could you check if it's acceptable for you?

gkurz replied:
> I've created a new commit. Could you check if it's acceptable for you?

Yes, thanks !

c3d commented on controllers/fg_handler.go:
```diff
@@ -82,7 +99,7 @@ func (r *KataConfigOpenShiftReconciler) processFeatureGates() error {
 	// Check which feature gates are enabled in the FG ConfigMap and
 	// perform the necessary actions
 	if r.kataConfig.Spec.EnablePeerPods {
-		if IsEnabled(fgStatus, ConfidentialFeatureGate) {
```
Even with a map `IsEnabled` could be a method so that you get `fgStatus.IsEnabled(Confidential)`, which I find more readable.

beraldoleal commented on config/peerpods/osc-configs-for-ds.yaml:
```diff
@@ -0,0 +1,585 @@
+  namespace: openshift-sandboxed-containers-operator
+#immutable: true
+data:
+  50-kata-remote: |
```
IIU, as it is, this DS mode would work only for peer-pods?

Unless I'm missing something, I feel this could have some overlap with @gkurz's work, on moving things to RPM side.

@gkurz is my understanding right? Would that work be ready for 1.10.2 ? Or post 1.10.2?

beraldoleal replied:
OK, I just got this answer, the agreement so far is that we keep it here for 1.10.2, but we should move this to the RPM for 1.11.

So we can close this thread.

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+			if updErr != nil {
+				return ctrl.Result{}, updErr
+			}
+			r.Log.Info("Kata PODs are present. Requeue for reconciliation ")
```
```suggestion
			r.Log.Info("Kata pods are present. Requeue for reconciliation ")
```

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+		return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, err
+	}
+	if err := controllerutil.SetControllerReference(r.kataConfig, kataUninstallDs, r.Scheme); err != nil {
+		r.Log.Error(err, "Failed setting ControllerReference for kata uninstallation DS")
```
```suggestion
		r.Log.Error(err, "Failed setting ControllerReference for kata uninstallation DaemonSet")
```

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+	r.setInProgressConditionToUninstalling()
+
+	// Delete the kata install ds if it exists
+	kataInstallDs, err := r.daemonSetForKataInstall(InstallKata)
```
Since DS is an acronym, suggest `kataInstallDS` or `kataInstallDaemonSet` at your preference. Note that `Ds` easily reads as _multiple `D`s_

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+
+	// Create kata uninstall ds if it doesn't exist
+	// Save whether it was just created or not
+	kataUninstallDs, err := r.daemonSetForKataInstall(UninstallKata)
```
Suggest `kataUninstallDaemonsSet` (or `kataUninstallDS`)

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+		return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, err
+	}
+
+	foundKataDs := &appsv1.DaemonSet{}
```
suggest `foundKataDaemonSet` or `foundKataDS`

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+			} else {
+				err = labelNodes(r.Client, kataSelector, map[string]string{kataDsInstallationLabel: string(KataDsWaitingToUninstall)})
+				if err != nil {
+					r.Log.Error(err, "could not add label waiting to uninstall to kata nodes")
```
```suggestion
					r.Log.Error(err, "could not add label while waiting to uninstall from kata nodes")
```

c3d commented on controllers/daemonset_reconcile.go:
```diff
@@ -144,10 +154,315 @@ func (r *KataConfigOpenShiftReconciler) daemonSetDeployment() (ctrl.Result, erro
+
+			kataSelector, err := r.getKataConfigNodeSelectorAsSelector()
+			if err != nil {
+				r.Log.Error(err, "could not get kata node selector, skipping waiting to uninstall labeling")
```
Sorry, I really have trouble parsing this error. What does "skipping waiting" mean in that context? Do you mean "skipping" waiting or "skipping to some later step", or "waiting for (labelling)", or some combination of the above?

```suggestion
				r.Log.Error(err, "could not get kata node selector, will skip labelling of nodes to uninstall")
```

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -2556,27 +2720,6 @@ func (r *KataConfigOpenShiftReconciler) disablePeerPods() error {
-		return err
-	}
-
 	return nil
```
Appreciate the effort into joining them, thanks.

However, this commit its changing parts that you introduced in this PR. It should be squashed with the appropriate commits.

beraldoleal reviewed (commented):
Hi @Pacho20, thanks again for this work.

The comments shows the importance of your work, This is a big feature and you’re making solid progress. Keep going.

Pacho20 committed:
daemonset-deployment: add utility functions
Introduced a helper function to fetch component images based on the cluster version.
The function depends on github.com/openshift/oc, which has been added to go.mod.
Also added a function to label nodes.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

beraldoleal commented on controllers/daemonset_reconcile.go:
```diff
@@ -0,0 +1,1085 @@
+	return nil
+}
+
+func (r *KataConfigOpenShiftReconciler) GetExtensionImage() (string, error) {
```
I just missed the reason we have `GetExentionImage()` and `GetCliImage()` methods. Can't we avoid both methods and calling directly the `GetImageForComponent` ?

Pacho20 replied:
We probably don't need the` GetCliImage()` function. I initially created both functions to allow for additional logic if needed. As you can see, `GetExtensionImage()` has a `TODO` comment because I wasn't sure whether this is the only extension image we need. Other extension images exist for different distros. If RHEL and CoreOS are sufficient, I can remove both functions and call `GetImageForComponent()` directly.

beraldoleal replied:
Please, call directly.

Pacho20 replied:
Done.

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -165,7 +167,17 @@ func (r *KataConfigOpenShiftReconciler) Reconcile(ctx context.Context, req ctrl.
+			case MachineConfigMode:
+				res, err = r.processKataConfigDeleteRequest()
+			case DaemonSetMode:
+				res, err = r.processKataConfigDeleteRequestDaemonSet()
```
What about the `DaemonSetFallback` ?

gcoon151 replied:
When the FG is set to DaemonSetFallback, this field is automatically set (L87 deployment_mode_handler.go) to either MachineConfigMode or DaemonSetMode depending on what is there. What's the suggestion?

beraldoleal replied:
I think I got it now. thanks. closing the thread.

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -181,7 +193,18 @@ func (r *KataConfigOpenShiftReconciler) Reconcile(ctx context.Context, req ctrl.
+		case MachineConfigMode:
+			res, err = r.processKataConfigInstallRequest()
+		case DaemonSetMode:
+			res, err = r.processKataConfigInstallRequestDaemonSet()
```
Same here, what about the `DaemonSetFallback`?

gcoon151 replied:
When the FG is set to DaemonSetFallback, this field is automatically set (L87 deployment_mode_handler.go) to either MachineConfigMode or DaemonSetMode depending on what is there. What's the suggestion?

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -1006,18 +1031,9 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
-			r.Log.Info("Kata PODs are present. Requeue for reconciliation ")
-			return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, err
-		}
+	res, err := r.checkDeletionEligibility()
```
Nice move!

gkurz replied:
Yes and it would be even nicer in a preparatory patch. There are far too many changes hitting `controllers/openshift_controller.go`... this makes the PR a lot harder to review.

beraldoleal commented on controllers/openshift_controller.go:
```diff
@@ -2489,6 +2576,47 @@ func (r *KataConfigOpenShiftReconciler) disablePeerPods() error {
+		err = r.deletePeedPodsConfigDaemonSet()
+		if err != nil {
+			return err
+		}
```
What about the fallback mode here ?

gcoon151 replied:
same as above. This is handled in deployment_mode_handler.go

gkurz commented on controllers/deployment_mode_handler.go:
```diff
@@ -0,0 +1,123 @@
+
+	// Attempt to GET any MachineConfig to verify if the GVK is known.
+	// If the CRD isn’t installed, it will return a NoMatchError.
+	err := r.Client.Get(context.Background(), client.ObjectKey{Name: machineConfigName}, mc)
```
This is the only sensible user for `machineConfigName`. No deep to define a constant for that. Let's use the literal string `"dummy-machine-config"` directly here.

Pacho20 replied:
Done

Pacho20 committed:
daemonset deployment: replace feature gate map with struct

Prepare feature gate for multi-value support by replacing map[string]bool with struct

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 committed:
daemonset deployment: add feature gate and detection for DaemonSet mode

Introduce a new feature gate to support running without the Machine Config Operator (MCO),
using a DaemonSet-based deployment instead. This commit adds only the feature gate
definition and detection logic.

The feature gate supports three options, which determine whether DaemonSet mode is activated.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

beraldoleal commented:
Some required test are failing, but might be because of the conflict. Could you rebase, please?

beraldoleal commented on scripts/kata-install/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+ADD osc-kata-install.sh osc-configs-script.sh osc-log-level.sh lib.sh /scripts/
+
+RUN curl -sSL "https://github.com/opencontainers/umoci/releases/download/v0.4.7/umoci.amd64" -o "/usr/local/bin/umoci" &&\
```
Future issue: This will be an issue for 1.11.0 (when we migrate this to Konflux). We need to make the builds hermetic.

gkurz replied:
It is possible to ask hermeto to prefetch arbitrary files for hermetic builds.

https://github.com/hermetoproject/hermeto/blob/main/docs/generic.md

gkurz commented:
/test ci-index-openshift-sandboxed-containers-operator-bundle

spotlesstofu reviewed (approved):
We can merge

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -213,9 +234,11 @@ func (r *KataConfigOpenShiftReconciler) Reconcile(ctx context.Context, req ctrl.
-		err = r.processLogLevel(r.kataConfig.Spec.LogLevel)
-		if err != nil {
-			res = ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}
+		if r.DeploymentMode == MachineConfigMode {
```
Can you clarify why this isn't wanted in `DaemonConfigMode` ?

Pacho20 replied:
The processLogLevel method creates a ContainerRuntimeConfig managed by the MCO. At the end of the day, it just results in a single file being created in CRI-O's config folder. I added this file creation step to the installation DaemonSet's script so no need to call this method for the DaemonSet mode.

gkurz replied:
You're right that the current implementation of `processLogLevel()` does MCO stuff that isn't suitable for your needs but the DaemonSet code still requires and processes the log level, right ?

Ideally we should then have a `processLogLevelDaemonSet()` function to do that or alternatively call the original `processLogLevel()` from another location that isn't MCO specific.

This might generate too many changes at this stage though so I'm fine to keep the `if` for now. Please add a `// FIXME` comment that this should be revisited later.

Pacho20 replied:
I wanted to create a separate DaemonSet to follow the reconcile logic, but creating one just for a single file seemed excessive. The operator already creates at least four different DaemonSets. If that's not a concern, we could introduce a processLogLevelDaemonSet() function in a future iteration down the road.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1181,15 +1195,9 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
 
-	ds := r.processDaemonsetForMonitor()
-	err = r.Client.Delete(context.TODO(), ds)
+	err = r.deleteDaemonsetForMonitor()
```
Please refactor the code in a separate commit.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1199,74 +1207,119 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
-
-		case RequeueNeeded:
-			r.setInProgressConditionToPodVMImageDeleting()
+		err := r.deletePodVMImage()
```
Please refactor the code in a separate commit.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1199,74 +1207,119 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
-		case ImageDeletionStatusUnknown:
-			r.setInProgressConditionToPodVMImageDeletionUnknown()
-			return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, err
+	err = r.deleteScc()
```
Please refactor the code in a separate commit.

gkurz reviewed (changes requested):
Hi @Pacho20 !

Since this affects the common code quite a lot, I want an easy review where I'm confident I can spot errors. The last commit mixes a lot of refactoring with the functional changes for the `DaemonSetMode`. Please move all the refactoring to preparatory commits.

gkurz commented:
> We can merge

I'd really prefer the last commit to be broken down before we merge.

Pacho20 committed:
daemonset deployment: add mvp for DaemonSet mode
This commit builds on the previously introduced
feature gate for DaemonSet-based deployment.
Adds support for deploying and managing the Kata and CAA installation using DaemonSets
without relying on the Machine Config Operator (MCO).

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 commented:
> > We can merge
> 
> I'd really prefer the last commit to be broken down before we merge.

@gkurz I've created a separate commit for the refactoring part. I hope this is what you had in mind.

Pacho20 committed:
daemonset deployment: introduce checkDeletionEligibility

Refactored logic that prevents deletion of KataConfig when KataPods are present into the checkDeletionEligibility method.

This method will be shared between MachineConfig and DaemonSet deployment modes.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 committed:
daemonset deployment: introduce deleteScc

Refactored logic for deleting the SCC into a new method.
This method will be shared between MachineConfig and DaemonSet deployment modes.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 committed:
daemonset deployment: introduce deletePeerPodsMC

Refactored code to introduce deletePeerPodsMC for handling deletion of PeerPods MachineConfig.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 committed:
daemonset deployment: introduce deleteDaemonSetForMonitor

Refactored the logic for deleting the monitor DaemonSet into a new method.
This method will be shared between MachineConfig and DaemonSet deployment modes.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1184,51 +1184,10 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
-		case ImageDeletionStatusUnknown:
-			r.setInProgressConditionToPodVMImageDeletionUnknown()
+		err := r.deletePodVMImage()
+		if err != nil {
```
We can get there from the `RequeueNeeded` branch in `deletePodVMImage()` with `err` being `nil` and we won't requeue... `deletePodVMImage()` should return `(*ctrl.Result, error)` and you should test either one being not `nil`.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1341,120 +1375,11 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
-		r.Log.Info("create Scc")
-		err = r.createScc()
+		err = r.postKataInstallation()
 		if err != nil {
```
Same issue as with `deletePodVMImage()`.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1184,51 +1184,13 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
+		if res != nil {
+			return *res, err
+		}
+		if err != nil {
```
This cannot happen AFAICT : all paths that  could return an error also return a reschedule request.

Pacho20 replied:
Yes, that's true. It was more of a safety measure in case someone changes the methods. What would you suggest instead?

gkurz replied:
I don't have an alternative to provide at this stage. We should carefully review all these return branches first.
So let's keep the safety measure around with a `// FIXME : dead code, revisit deletePodVMImage() return paths` comment.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1344,120 +1378,13 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
-
-		foundDs := &appsv1.DaemonSet{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: ds.Name, Namespace: ds.Namespace}, foundDs)
 		if err != nil {
```
Same remark as with `deletePodVMImage()` : all return paths of `postKataInstallation()` that return an error also return a reschedule request.

gkurz replied:
and `// FIXME : dead code. Revisit ...` comment as well here.

Pacho20 committed:
daemonset deployment: introduce deletePodVMImage

Refactored logic for deleting the PodVM image into a new method.
This method will be shared between MachineConfig and DaemonSet deployment modes.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

Pacho20 committed:
daemonset deployment: introduce createDaemonsetForMonitor and postKataInstallation

Refactored code to introduce createDaemonsetForMonitor for handling the creation of the Monitor DaemonSet.

Also refactored code to introduce postKataInstallation, which performs all necessary setup steps after Kata is successfully installed.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

beraldoleal reviewed (approved):
After addressing @gkurz's comments, LGTM.

Thanks again, @Pacho20. Well done.

gkurz commented on scripts/kata-install/configuration-remote.toml:
Yet another instance of `scripts/kata-install/configuration-remote.toml` in the tree... fortunately we're working on moving the kata and CRI-O conf files to the RPM for OSC 1.11 since they are static.

Pacho20 replied:
This is needed for the 1.10 release, but I’m looking forward to deleting these files and updating the DaemonSet for 1.11.

gkurz commented on scripts/kata-install/50-kata-remote:
As said in another comment, we'll get this file from the RPM in the future.

gkurz reviewed (approved):
/lgtm

Thanks for the very good work and patience @Pacho20 !

gkurz commented:
/ok-to-test

gkurz commented:
/ok-to-test

littlejawa commented:
/retest

gkurz commented:
Fixes https://issues.redhat.com/browse/KATA-3962

gkurz commented:
Fixes https://issues.redhat.com/browse/KATA-3963
